File: buildings/simulate2R2C.py
"""
%% introduction
This script models a building as a 2R2C thermal circuit and simulates its
dynamics both exactly and using a two-timing approximation.

Author: Kevin J. Kircher, Purdue University, 2025

Please make sure the following functions are available:
    importWeather.py (imports weather data)
    importElectricity.py (imports electricity load data)
    plotResults.py (plots simulation results)
    perfectTrackingControl.py (students fill this in)
    thermostaticControl.py (students fill this in)
"""

# ==============================================================================
# Required imports
# ==============================================================================
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from perfectTrackingError import perfect_tracking_control
from thermostaticControl import thermostatic_control
from plotRCresults import plot_rc_results
from scipy.linalg import expm
from importWeather import import_weather
from importElectricty import import_electricity


# ==============================================================================
# Graphics settings
# ==============================================================================

plt.rc('font', size=20)
plt.rc('axes', titlesize=24, labelsize=20)
plt.rc('xtick', labelsize=20)
plt.rc('ytick', labelsize=20)
plt.rc('legend', fontsize=12)
plt.rc('figure', titlesize=20)
plt.rc('lines', linewidth=3)


# ==============================================================================
# Parameters
# ==============================================================================
# Geometry
Af = 200  # floor area, m^2
N = 2  # number of stories

# Thermal capacitances
C = 0.0125 * Af  # air thermal capacitance, kWh/C
Cm = 12 * C  # mass thermal capacitance, kWh/C

# Thermal resistances
R = 1 / (0.016 * np.sqrt(N * Af))  # indoor-outdoor thermal resistance, C/kW
Rm = R / 6  # indoor-mass thermal resistance, C/kW

# Timing
dt = 0.25 #time step, h
t_span = pd.date_range("2022-12-22", "2022-12-27", freq="15T") #time span as datetime
K = len(t_span) # number of time steps
t = np.arange(0, K * dt, dt) # time span, h

# Heater capacity constraints
qcMin = np.zeros(K) #minimum HVAC thermal power, kW
qcMax = 14 * np.ones(K) #maximum HVAC thermal power, kW

# ==============================================================================
# 2R2C model
# ==============================================================================
# Continuous-time dynamics matrices
Ac = np.zeros((2, 2))  # continuous-time dynamics matrix
Bc = np.zeros((2, 1))  # continuous-time input matrix

# your
#
# code
#
# here

# Discrete-time dynamics matrices
# your code
# here


# ==============================================================================
# Input signals
# ==============================================================================
# Weather data import
weather_file = 'west-lafayette-2022-weather.csv'
Tout, I = import_weather(weather_file, t_span)

# Electricity data import
electricity_file = 'MFRED-2019-NYC-Apartments-Electricity-Data.csv'
t_elec = t_span.map(lambda x: x.replace(year=2019))
plug_powers = import_electricity(electricity_file, t_elec)

# Exogenous thermal power
qe = (plug_powers[:, np.random.randint(plug_powers.shape[1])] +  # from plugged-in devices
      0.19 * np.sqrt(N * Af) * I +  # from the sun
      1 + (0.5 / 3) * np.random.randn(K))  # from everything else
# Disturbance
w = qe + Tout / R  # kW

# Indoor temperature setpoint
Tset = 21 * np.ones(len(t) + 1) #day indoor temperature setpoint, C
Tset[:len(t)][np.mod(t, 24) < 6] = 18.5 #night setpoint, C
Tset[:len(t)][np.mod(t, 24) > 22] = 18.5 #night setpoint, C

# %% input signal plot
plt.figure(1)
plt.clf()

# First subplot (Outdoor Temperature)
ax1 = plt.subplot(2, 1, 1)
plt.plot(t_span, Tout, 'k')
plt.ylabel('Outdoor \n temperature \n (°C)', fontsize=16)
plt.xticks(fontsize=14)
plt.yticks(fontsize=14)
plt.grid(True, linestyle='--', alpha=0.7)
ax1.xaxis.set_major_formatter(mdates.DateFormatter('%b %d %Y'))

# Second subplot (Exogenous Thermal Power)
ax2 = plt.subplot(2, 1, 2)
plt.plot(t_span, qe, 'k')
plt.ylabel('Exogenous \n thermal power \n (kW)', fontsize=16)
plt.xticks(fontsize=14)
plt.yticks(fontsize=14)
plt.grid(True, linestyle='--', alpha=0.7)
ax2.xaxis.set_major_formatter(mdates.DateFormatter('%b %d %Y'))

plt.tight_layout()
plt.show()

# ==============================================================================

# %% (near-) perfect tracking simulation
# initial state
T0 = np.array([Tset[0], np.mean(Tset)])  # initial air and mass temperature, C

# simulation
T1, qc1 = perfect_tracking_control(A, B, w, T0, Tset, qcMin, qcMax)

# plot results
plot_rc_results(t, Tset[:-1], T1[0, :-1], T1[1, :-1], qc1, 2)

# ==============================================================================

# %% thermostatic control simulation
# simulation
dT = 0.5  # thermostat deadband halfwidth, C
T2, qc2 = thermostatic_control(A, B, w, T0, Tset, qcMin, qcMax, dT)

# plot results
plot_rc_results(t, Tset[:-1], T2[0, :-1], T2[1, :-1], qc2, 3)

File: buildings/simulate2R2C.m
%% introduction
% This script models a building as a 2R2C thermal circuit and simulates its
% dynamics both exactly and using a two-timing approximation.
%
% Kevin J. Kircher, Purdue University, 2024
%
% Please make sure the following functions are on your Matlab path:
%   importWeather.m (imports weather data)
%   importElectricity.m (imports electricity load data)
%   plotResults.m (plots simulation results)
%   fullStairs.m (stairstep plots)
%   perfectTrackingControl.m (students fill this in)
%   thermostaticControl.m (students fill this in)

%% graphics settings
% I recommend adding these to a startup.m file in your Matlab root 
% directory. Matlab runs that file at startup, effectively turning these 
% settings into new defaults.

% set LaTeX as the default text interpreter
set(0,'DefaultTextInterpreter','latex');
set(0,'DefaultLegendInterpreter','latex');
set(0,'DefaultAxesTickLabelInterpreter','latex');

% set the default font size for axes ticks/labels and figure titles
set(0,'DefaultAxesFontSize',24);

% set the default text size for figures and legends
set(0,'DefaultTextFontSize',24);
set(0,'DefaultLegendFontSizeMode','manual')
set(0,'DefaultLegendFontSize',24)

% set the default line width for plots and stairstep plots
set(0,'DefaultLineLineWidth',2.5);
set(0,'DefaultStairLineWidth',2.5);

%% parameters
% geometry
Af = 200; % floor area, m^2
N = 2; % number of stories

% thermal capacitances
C = 0.0125*Af; % air thermal capacitance, kWh/C
Cm = 12*C; % mass thermal capacitance, kWh/C

% thermal resistances
R = 1./(0.016*sqrt(N*Af)); % indoor-outdoor thermal resistance, C/kW
Rm = R/6; % indoor-mass thermal resistance, C/kW

% timing
dt = 0.25; % time step, h
tSpan = (datetime(2022,12,22,0,0,0):hours(dt):datetime(2022,12,26,23,0,0))'; % time span as datetime
K = length(tSpan); % number of time steps
t = (0:dt:K*dt)'; % time span as float, h

% heater capacity constraings
qcMin = zeros(K,1); % minimum HVAC thermal power, kW
qcMax = 14*ones(K,1); % maximum HVAC thermal power, kW

%% 2R2C system matrices
% continuous-time system matrices
Ac = zeros(2,2); % continuous-time dynamics matrix
Bc = zeros(2,1); % continuous-time input matrix
% your
% 
% code
% 
% here

% discrete-time system matrices
% your code
% here

%% input signals
% weather data import
weatherFile = 'west-lafayette-2022-weather.csv';
[Tout,I,~,~,~] = importWeather(weatherFile,tSpan);

% electricity data import
electricityFile = 'MFRED-2019-NYC-Apartments-Electricity-Data.csv';
tElec = tSpan; tElec.Year = 2019;
plugPowers = importElectricity(electricityFile,tElec);

% exogenous thermal power
lambda = 0.25; % glazing ratio
c = 0.4; % solar heat gain coefficient
qe = plugPowers(:,randi(size(plugPowers,2))) ... % from plugged-in devices
    + 4.8*c*lambda*sqrt(N*Af*I) ... % from the sun
    + 1 + (0.5/3)*randn(K,1); % from everything else

% disturbance
w = qe + Tout/R; % kW

% indoor temperature setpoint
Tset = 21*ones(K+1,1); % day indoor temperature setpoint, C
Tset(mod(t,24) < 6 | mod(t,24) > 22) = 18.5; % night setpoint, C

% input signal plot
figure(1), clf
subplot(2,1,1), plot(tSpan,Tout,'k')
ylabel({'Outdoor','temperature','($^\circ$C)'})
subplot(2,1,2), plot(tSpan,qe,'k')
ylabel({'Exogenous','thermal power','(kW)'})

%% (near-) perfect tracking simulation
% initial state
T0 = [Tset(1); % initial air temperature, C
    mean(Tset)]; % initial mass temperature, C

% simulation
[T1,qc1] = perfectTrackingControl(A,B,w,T0,Tset,qcMin,qcMax);

% plot results
plotRCresults(t,Tset,T1(1,:),T1(2,:),qc1,2)

%% thermostatic control simulation
% simulation
dT = 0.5; % thermostat deadband halfwidth, C
[T2,qc2] = thermostaticControl(A,B,w,T0,Tset,qcMin,qcMax,dT);

% plot results
plotRCresults(t,Tset,T2(1,:),T2(2,:),qc2,3)


















